src/websocket.rs

```diff
@@ -0,0 +1,372 @@
+#![allow(dead_code)]
+
+use crate::error::HeliusError;
+use crate::types::{RpcTransactionsConfig, TransactionNotification};
+use futures_util::{
+    future::{ready, BoxFuture, FutureExt},
+    sink::SinkExt,
+    stream::{BoxStream, StreamExt},
+};
+use serde::de::DeserializeOwned;
+use serde_json::{json, Map, Value};
+use solana_account_decoder::UiAccount;
+use solana_rpc_client_api::config::{RpcAccountInfoConfig, RpcProgramAccountsConfig};
+use solana_rpc_client_api::filter::maybe_map_filters;
+use solana_rpc_client_api::response::RpcKeyedAccount;
+use solana_rpc_client_api::{
+    error_object::RpcErrorObject,
+    response::{Response as RpcResponse, RpcVersionInfo},
+};
+use solana_sdk::pubkey::Pubkey;
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use tokio::{
+    net::TcpStream,
+    sync::{mpsc, oneshot, RwLock},
+    task::JoinHandle,
+    time::{sleep, Duration},
+};
+use tokio_stream::wrappers::UnboundedReceiverStream;
+use tokio_tungstenite::{
+    connect_async,
+    tungstenite::{
+        protocol::frame::{coding::CloseCode, CloseFrame},
+        Message,
+    },
+    MaybeTlsStream, WebSocketStream,
+};
+use url::Url;
+
+pub const ENHANCED_WEBSOCKET_URL: &str = "wss://atlas-mainnet.helius-rpc.com?api-key=";
+
+pub type EnhancedWebsocketResult<T = ()> = Result<T, HeliusError>;
+
+type UnsubscribeFn = Box<dyn FnOnce() -> BoxFuture<'static, ()> + Send>;
+type SubscribeResponseMsg = Result<(mpsc::UnboundedReceiver<Value>, UnsubscribeFn), HeliusError>;
+type SubscribeRequestMsg = (String, Value, oneshot::Sender<SubscribeResponseMsg>);
+type SubscribeResult<'a, T> = EnhancedWebsocketResult<(BoxStream<'a, T>, UnsubscribeFn)>;
+type RequestMsg = (String, Value, oneshot::Sender<Result<Value, HeliusError>>);
+
+/// A client for subscribing to messages from the RPC server.
+///
+/// Forked from Solana's EnhancedWebsocket.
+#[derive(Debug)]
+pub struct EnhancedWebsocket {
+    subscribe_sender: mpsc::UnboundedSender<SubscribeRequestMsg>,
+    request_sender: mpsc::UnboundedSender<RequestMsg>,
+    shutdown_sender: oneshot::Sender<()>,
+    node_version: RwLock<Option<semver::Version>>,
+    ws: JoinHandle<EnhancedWebsocketResult>,
+}
+
+impl EnhancedWebsocket {
```

Comment:
More fleshed-out documentation for these new methods would be greatly appreciated